### PR TITLE
Switch adoptopenjdk cask to use pkg

### PR DESCRIPTION
**Pull request by gdams**
This PR swaps the `adoptopenjdk` cask to use the notarized `pkg` installer files to avoid macOS signature errors as detailed in https://github.com/AdoptOpenJDK/homebrew-openjdk/issues/267

After making all changes to the cask:

- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` reports no offenses.
- [x] The commit message includes the cask’s name and version.
- [x] The submission is for [a stable version](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#finding-a-home-for-your-cask) or [documented exception](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#but-there-is-no-stable-version).

### CODE REVIEW

File: Casks/adoptopenjdk.rb [hunk 1]
```diff
@@ -1,14 +1,16 @@
 cask 'adoptopenjdk' do
   version '13.0.1,9'
-  sha256 '9c82de98ce9bc2353bcf314d85366c9a2c572db034e10a71aa47e804e13748c1'
+  sha256 '29f1c87b3ecca38c50ea016657368af0a6a28ade5ba4150bd147ad91c524ea22'
 
   # github.com/AdoptOpenJDK was verified as official when first introduced to the cask
-  url "https://github.com/AdoptOpenJDK/openjdk#{version.major}-binaries/releases/download/jdk-#{version.before_comma}%2B#{version.after_comma}/OpenJDK#{version.major}U-jdk_x64_mac_hotspot_#{version.before_comma}_#{version.after_comma}.tar.gz"
+  url 'https://github.com/AdoptOpenJDK/openjdk13-binaries/releases/download/jdk-13.0.1%2B9/OpenJDK13U-jdk_x64_mac_hotspot_13.0.1_9.pkg'
```
Comment (vitorgalvao) [inline Casks/adoptopenjdk.rb]: Should still use the interpolation.

File: Casks/adoptopenjdk.rb [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
   appcast "https://github.com/AdoptOpenJDK/openjdk#{version.major}-binaries/releases/latest"
   name 'AdoptOpenJDK Java Development Kit'
   homepage 'https://adoptopenjdk.net/'
 
-  artifact "jdk-#{version.before_comma}+#{version.after_comma}", target: "/Library/Java/JavaVirtualMachines/adoptopenjdk-#{version.before_comma}.jdk"
+  pkg 'OpenJDK13U-jdk_x64_mac_hotspot_13.0.1_9.pkg'
```
Comment (vitorgalvao) [inline Casks/adoptopenjdk.rb]: Should still use the interpolation.

**Review by vitorgalvao — CHANGES_REQUESTED** [review]

**Comment by vitorgalvao** [comment]
I prefer the `artifact` approach since a `pkg` can do anything to the system, but I’ll leave the decision to @reitermarkus, since he introduced the `artifact` method.

I’ll note that the current problem would be fixed by he solutions proposed in https://github.com/Homebrew/homebrew-cask/issues/70798.

**Comment by zshift** [comment]
Would it be possible to add me as a collaborator to this PR?

**Comment by JuanitoAlimana** [comment]
No

**Comment by vitorgalvao** [comment]
@JuanitoAlimana Please stop randomly replying to posts from other users. The other ones were confusing, but this one might be interpreted as rude.

I’ll give you the benefit of the doubt for now and assume you’re still understanding Github, but if you keep it up we’ll have to block you from the repo. I’d rather not have that happen.

**Comment by JuanitoAlimana** [comment]
[Sorry, but who is rude is your, when I unsubscribre and you keep all
this...]

File: Casks/adoptopenjdk.rb (hunk 1 above)
Comment (zshift) [inline Casks/adoptopenjdk.rb]:
```suggestion
  url 'https://github.com/AdoptOpenJDK/openjdk#{version.major}-binaries/releases/download/jdk-#{version.before_comma}%2B#{version.after_comma}/OpenJDK#{version.major}U-jdk_x64_mac_hotspot_#{version.before_comma}_#{version.after_comma}.pkg'
```

File: Casks/adoptopenjdk.rb (hunk 2 above)
Comment (zshift) [inline Casks/adoptopenjdk.rb]:
```suggestion
  pkg 'OpenJDK#{version.major}U-jdk_x64_mac_hotspot_#{version.before_comma}_#{version.after_comma}.pkg'
```

**Comment by JuanitoAlimana** [comment]
Fuck off!!!

Sent from iPhone

> El 10/02/2020, a la(s) 11:47 a. m., Peter David Faria <notifications@github.com> escribió:
> 

﻿
@zshift commented on this pull request.

In Casks/adoptopenjdk.rb:

>  
   # github.com/AdoptOpenJDK was verified as official when first introduced to the cask
-  url "https://github.com/AdoptOpenJDK/openjdk#{version.major}-binaries/releases/download/jdk-#{version.before_comma}%2B#{version.after_comma}/OpenJDK#{version.major}U-jdk_x64_mac_hotspot_#{version.before_comma}_#{version.after_comma}.tar.gz"
+  url 'https://github.com/AdoptOpenJDK/openjdk13-binaries/releases/download/jdk-13.0.1%2B9/OpenJDK13U-jdk_x64_mac_hotspot_13.0.1_9.pkg'
⬇️ Suggested change
-  url 'https://github.com/AdoptOpenJDK/openjdk13-binaries/releases/download/jdk-13.0.1%2B9/OpenJDK13U-jdk_x64_mac_hotspot_13.0.1_9.pkg'
+  url 'https://github.com/AdoptOpenJDK/openjdk#{version.major}-binaries/releases/download/jdk-#{version.before_comma}%2B#{version.after_comma}/OpenJDK#{version.major}U-jdk_x64_mac_hotspot_#{version.before_comma}_#{version.after_comma}.pkg'

**Comment by vitorgalvao** [comment]
Apologies to everyone else on the issue. The disruptive user has now been blocked and their comments hidden.

**Comment by gdams** [comment]
This bug has been raised yet again, https://github.com/AdoptOpenJDK/homebrew-openjdk/issues/329.

Please consider switching to our pkg's.

See https://github.com/AdoptOpenJDK/homebrew-openjdk/blob/master/Casks/adoptopenjdk14.rb for the pkg equivalent

**Comment by vitorgalvao** [comment]
@gdams Considering there’s now an [`openjdk` formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/openjdk.rb), is there a point to `adoptopenjdk` as a cask?

**Comment by gdams** [comment]
>@gdams Considering there’s now an openjdk formula, is there a point to adoptopenjdk as a cask?

@vitorgalvao that cask will also experience the same errors as it's not been notarized. If anything we should migrate to AdoptOpenJDK as the main provider because we do provide notarized installers. AdoptOpenJDK is also the second most used OpenJDK behind Oracle's Java so would be the obvious default.

**Comment by vitorgalvao** [comment]
> AdoptOpenJDK is also the second most used OpenJDK behind Oracle's Java so would be the obvious default.

Granted, I don’t know much about java, but there’s nothing obvious about “we’re the second in popularity, so we should be the main offering”.

> that cask will also experience the same errors as it's not been notarized.

Not a cask, a formula.

And it doesn’t experience the same issue because it’s not quarantined, hence notarisation checks don’t apply. Those only occur in quarantined software, which is a flag the downloading software must add. `curl`, which we use, doesn’t add the flag. It’s only added in casks to great pains and we [are considering backtracking on](https://github.com/Homebrew/homebrew-cask/issues/70798),

In the cask we have for `adopopenjdk`, [if it’s installed with `--no-quarantine`, the issue goes away](https://github.com/Homebrew/homebrew-cask/issues/79038#issuecomment-602686706). 

So if the point of `adoptopenjdk` is that it’s notarised and notarisation isn’t an issue, I repeat my question.

**Comment by aahlenst** [comment]
@vitorgalvao Disclosure: I'm an AdoptOpenJDK member. 

[OpenJDK](https://openjdk.java.net) is the home of Java development. But it is not the home of binary distributions (it's a little bit like Linux with kernel.org and Red Hat etc.). You can only get binary distributions of the newest JDK on https://jdk.java.net. Currently, this is JDK 14. For all other versions, you have to turn to someone else. Some of the options: [AdoptOpenJDK](https://adoptopenjdk.net/), [Amazon](https://aws.amazon.com/corretto/), [Azul](https://www.azul.com), [BellSoft](https://bell-sw.com), [SAP](https://sap.github.io/SapMachine/). Oracle provides binary distributions, too, but they are only free for personal use. All the others I mentioned are free without restrictions. Now, what most users need is not the newest JDK. Instead, they need to be able to choose from multiple versions. For example, Android development is currently only possible with JDK 8. The minimum offering should include all JDKs with long term support plus the newest one. Currently, the selection would be 8, 11 and 14. Ideally, all currently supported JDKs should be available. At the moment, those would be 8, 11, 13 and 14. 

The second problem I see: The openjdk@11 formula is built from source without putting it through a test suite. This is frowned upon in the Java community. At least AdoptOpenJDK, Azul, BellSoft and SAP test their binaries extensively. For Amazon, AdoptOpenJDK performs the testing.

With my developer hat on (and using Homebrew): The current offering of formulae is insufficient plus I don't like that openjdk@11 is built from source. Either there should be multiple formulae using tested, unrestricted binaries from a trusted upstream provider (be it AdoptOpenJDK, Azul or BellSoft) or Homebrew should only offer the newest JDK directly from OpenJDK and point to trusted 3rd party casks.

With my supporter hat on: Homebrew should make it possible to install JDKs that work out of the box without additional flags like --no-quarantine.

**Comment by SMillerDev** [comment]
> The second problem I see: The openjdk@11 formula is built from source without putting it through a test suite. This is frowned upon in the Java community. At least AdoptOpenJDK, Azul, BellSoft and SAP test their binaries extensively. For Amazon, AdoptOpenJDK performs the testing.

All formulae that depend on it will be tested for functionality on top of the test block being run. I don't imagine other package managers distribute Java versions differently?

**Comment by aahlenst** [comment]
@SMillerDev Some Linux distributions have their own builds that do not go through TCK or a comparable test suite. I‘ve seens issues like a broken jlink, something that is typically not tested by downstream applications. Java‘s surface area is very large and it is very easy to break things without testing. There is nothing wrong with building from source, but the result should be tested. If Homebrew wants to do that, you‘re welcome to use the test suite created by AdoptOpenJDK.

**Comment by MikeMcQuaid** [comment]
@aahlenst If you are the person who wants the test suite added: you are welcome to submit a PR to add it to the relevant formulae. 

Can you point to any bugs with our formulae as-is or do you just "not like it because it's built from source"?

---

As to this thread in general: I'm in favour of us continuing to ship the quarantined version as well as a formula.

**Comment by gdams** [comment]
> @aahlenst If you are the person who wants the test suite added: you are welcome to submit a PR to add it to the relevant formulae.

@MikeMcQuaid testing Java is a very long process, a typical AdoptOpenJDK release can take somewhere between 4-12 hours running tests. This is obviously something that would not make sense to add to a cask.

I have never heard of a Linux distro that doesn't at least run a small subset of tests before shipping openjdk. Homebrew is sending out the wrong impression if they are actively encouraging users to run an untested OpenJDK binary in production.

Developers typically want to run the same build of Java in development as they do in production. That's why so many people rely on AdoptOpenJDK on their dev machines.

**Comment by MikeMcQuaid** [comment]
> I have never heard of a Linux distro that doesn't at least run a small subset of tests before shipping openjdk. Homebrew is sending out the wrong impression if they are actively encouraging users to run an untested OpenJDK binary in production.

We run tests before shipping openjdk updates. We don't really appreciate insinuation that this is not the case, thanks.

> Developers typically want to run the same build of Java in development as they do in production.

People using an open-source package manager typically trust the people running that package manager to make the right decisions on how that software is packaged. People using an open-source package manager generally expect open source software to be built from source. 

I'm all for finding out ways for your project to reduce your support burden but this is getting a bit "we'll tell you how best to run Homebrew". We're running the package manager and you're building Java: can we assume that we maybe know a bit more about running a package manager and you know a bit more about building Java? Thanks :heart:

**Comment by gdams** [comment]
@MikeMcQuaid as you say homebrew is an open-source package manager and we love it! In no way do I want you to feel that I'm telling you how to run homebrew.

> We run tests before shipping openjdk updates.

Can you provide more insight into what tests are run by homebrew before shipping?

**Comment by MikeMcQuaid** [comment]
> Can you provide more insight into what tests are run by homebrew before shipping?

We run the integration tests of every Homebrew binary package that (recursively) uses the `openjdk` formula on any changes to that formula.

**Comment by aahlenst** [comment]
@MikeMcQuaid Questions were raised like why could something else be needed besides the `openjdk` formula and it was indicated that it's unclear why OpenJDK is different than other software. I tried to answer those questions from multiple perspectives in the most neutral way possible. If I did a bad job and you feel offended, I'm sorry. My intention wasn't to demand that you do things differently, merely to explain our perspective.

To come back to the actual topic: Homebrew offers an `adoptopenjdk` cask. If this cask is installed with `brew cask install adoptopenjdk`, `java` does not work  because of the quarantine flag. This confuses our common user base and causes support requests for Homebrew *and* AdoptOpenJDK. We offered a notarized PKG with hardened runtime support. This was rejected while it was accepted for Azul Zulu (no hard feelings here).

So: Can we work together to fix this? Or should we abandon the `adoptopenjdk` cask and only offer our tap?

**Comment by MikeMcQuaid** [comment]
> If I did a bad job and you feel offended, I'm sorry. My intention wasn't to demand that you do things differently, merely to explain our perspective.

Thanks ❤️. Text-based communication is nuanced and hard (particularly right now). I greatly appreciate the apology and absolutely no hard feelings. My communication was/is never perfect either.



> If this cask is installed with `brew cask install adoptopenjdk`, `java` does not work because of the quarantine flag.

This doesn't quite seem to be the case for all users. I'll defer to @vitorgalvao for more details here.



> Can we work together to fix this? Or should we abandon the `adoptopenjdk` cask and only offer our tap?

Yes, I'm sure we can figure something out. I defer to @homebrew/cask folks in general on this repo, though.

**Comment by MikeMcQuaid** [comment]
People following along: @gadams and I had a Zoom call to chat about this a bit and I'm going to talk to @vitorgalvao and other @Homebrew/cask folks before we post anything else in here. Thanks for your patience!

**Comment by vitorgalvao** [comment]
Apologies for having taken so long to sort this out. We’ve talked, and we’re in agreement that changing this to a `.pkg` is the way to go.

Thank you to the AdoptOpenJDK team for their patience and answering all our questions in detail.

Is there anything you’d like to change (e.g. update the version) before this gets merged? If not, feel free to ping me (@gdams or @aahlenst) and I’ll go ahead and merge.

**Comment by gdams** [comment]
@vitorgalvao thanks for the update, this should be good to go if you're happy to merge it?

Thanks

George

**Pull request merged**